File: netlogo-gui/src/main/prim/_nsum.java
// (C) Uri Wilensky. https://github.com/NetLogo/NetLogo

package org.nlogo.prim;

import org.nlogo.agent.AgentIterator;
import org.nlogo.agent.AgentSet;
import org.nlogo.agent.Patch;
import org.nlogo.agent.Turtle;
import org.nlogo.api.Dump;
import org.nlogo.core.I18N;
import org.nlogo.api.LogoException;
import org.nlogo.core.Syntax;
import org.nlogo.api.TypeNames;
import org.nlogo.nvm.Context;
import org.nlogo.nvm.RuntimePrimitiveException;
import org.nlogo.nvm.Reporter;

public final class _nsum extends Reporter {
  public int vn;

  public _nsum(int vn) {
    super();
    this.vn = vn;
  }

  @Override
  public String toString() {
    if (world == null) {
      return super.toString() + ":" + vn;
    }
    return super.toString() + ":" + world.patchesOwnNameAt(vn);
  }

  @Override
  public Object report(Context context) throws LogoException {
    return report_1(context);
  }

  public double report_1(Context context) throws LogoException {
    Patch patch;
    if (context.agent instanceof Turtle) {
      patch = ((Turtle) context.agent).getPatchHere();
    } else {
      patch = (Patch) context.agent;
    }
    double sum = 0;
    for (AgentIterator it = patch.getNeighbors().iterator(); it.hasNext();) {
      Object value = ((Patch) it.next()).getPatchVariable(vn);
      if (!(value instanceof Double)) {
        throw new RuntimePrimitiveException(context, this,
            I18N.errorsJ().getN("org.nlogo.prim.$common.noSumOfListWithNonNumbers",
                Dump.logoObject(value).toString(), TypeNames.name(value)));
      }

      sum += ((Double) value).doubleValue();
    }
    return validDouble(sum, context);
  }
}

File: netlogo-headless/src/main/prim/_nsum.scala
// (C) Uri Wilensky. https://github.com/NetLogo/NetLogo

package org.nlogo.prim

import org.nlogo.agent.{ Patch, Turtle }
import org.nlogo.api.{ Dump, TypeNames }
import org.nlogo.core.I18N
import org.nlogo.nvm.{ Context, Reporter }
import org.nlogo.nvm.RuntimePrimitiveException

class _nsum(vn: Int) extends Reporter {
  override def toString =
    super.toString + ":" +
      Option(world).map(_.patchesOwnNameAt(vn)).getOrElse(vn.toString)

  override def report(context: Context): java.lang.Double =
    Double.box(report_1(context))

  def report_1(context: Context): Double = {
    val patch = context.agent match {
      case t: Turtle => t.getPatchHere
      case p: Patch => p
      case a => throw new Exception(s"Unexpected agent: $a")
    }
    var sum = 0d
    val it = patch.getNeighbors.iterator
    while(it.hasNext)
      it.next().asInstanceOf[Patch].getPatchVariable(vn) match {
        case d: java.lang.Double =>
          sum += d.doubleValue
        case x =>
          throw new RuntimePrimitiveException(
            context, this, I18N.errors.getN(
              "org.nlogo.prim.$common.noSumOfListWithNonNumbers",
              Dump.logoObject(x).toString, TypeNames.name(x)))
      }
    validDouble(sum, context)
  }

}
